File: src/std/java/src/fan/std/RegexPeer.java
package fan.std;

import java.util.regex.Pattern;

import fan.sys.FanStr;
import fan.sys.List;
import fanx.interop.Interop;

public class RegexPeer {
	private Pattern pattern;

	public static RegexPeer make(Regex self) {
		return new RegexPeer();
	}

	void init(Regex self) {
		pattern = Pattern.compile(self.source);
	}

	public boolean matches(Regex self, String s) {
		return pattern.matcher(s).matches();
	}

	public RegexMatcher matcher(Regex self, String s) {
		return new RegexMatcher(pattern.matcher(s));
	}

	public List split(Regex self, String s) {
		return split(self, s, 0L);
	}

	public List split(Regex self, String s, long limit) {
		return Interop.toFanList(FanStr.type, pattern.split(s, (int) limit));
	}
}

File: src/std/js/RegexPeer.js
fan.std.RegexPeer = function(){}

fan.std.RegexPeer.make = function(self) {
  return new fan.std.RegexPeer();
}

fan.std.RegexPeer.prototype.init = function(self) {
    self.m_regexp = new RegExp(self.m_source);
}

fan.std.RegexPeer.prototype.matches = function(self, s)
{
  return this.matcher(s).matches();
}

fan.std.RegexPeer.prototype.matcher = function(self, s)
{
  return new fan.std.RegexMatcher(self.m_regexp, self.m_source, s);
}

fan.std.RegexPeer.prototype.split = function(self, s, limit)
{
  if (limit === undefined) limit = 0;

  if (limit === 1)
    return fan.sys.List.makeFromJs(fan.sys.Str.$type, [s]);

  var array = [];
  var re = self.m_regexp;
  while (true)
  {
    var m = s.match(re);
    if (m == null || (limit != 0 && array.length == limit -1))
    {
      array.push(s);
      break;
    }
    array.push(s.substring(0, m.index));
    s = s.substring(m.index + m[0].length);
  }
  // remove trailing empty strings
  if (limit == 0)
  {
    while (array[array.length-1] == "") { array.pop(); }
  }
  return fan.sys.List.makeFromJs(fan.sys.Str.$type, array);
}

